apache/incubator-openwhisk
# Add PHP 7.1 as a kind

This adds a new `php:7.1` kind to allow PHP actions like this:

```php
<?php
function main(array $args) : array
{
    $name = $args["name"] ?? "stranger";
    $greeting = "Hello $name!";
    return ["greeting" => $greeting];
}
```

The PHP runtime is based on the official `php:7.1-alpine` one with the following extensions added:

- opcache
- mysqli
- pdo_mysql
- pdo_pgsql
- intl
- bcmath
- zip
- gd
- soap

Two Composer packages are also included:

- `guzzlehttp/guzzle` - A very popular HTTP client
- `ramsey/uuid` - The de facto UUID generator

Suggestions welcome on what should be included in these lists (and of course on the code itself!)


## How to test

We need a working local installation. The easiest way to do this is with Vagrant, so that's what this set of steps uses.

1. Check out the source code 

        $ git clone https://github.com/apache/incubator-openwhisk.git
        $ cd incubator-openwhisk


2. Now change to my branch

        $ git remote add akrabat https://github.com/akrabat/openwhisk.git
        $ git fetch akrabat
        $ git checkout php-kind

3. Run up the Vagrant box

        $ cd tools/vagrant
        $ vagrant up

    This takes a while!
   
    I'm not sure what Vagrant is required, but I'm using Vagrant 1.9.1 with VirtualBox 5.1.18. I'm not aware of any other requirements.

    You should see output at the end like:

        ==> default: ++ wsk action invoke /whisk.system/utils/echo -p message hello --result
        ==> default: {
        ==> default:     "message": "hello"
        ==> default: }
        ==> default: +++ date
        ==> default: ++ echo 'Sun Jul 23 14:28:31 UTC 2017: build-deploy-end'

4. Build the PHP Docker container

        $ vagrant ssh
        $ cd openwhisk
        $ ./gradlew :core:php7.1Action:distDocker -PdockerImagePrefix=openwhisk

    This takes a little while while it compiles various PHP bits and bobs.

4. Testing a PHP action

    The vagrant box has a compatible `wsk` in it, so we can use it:

        $ cd ~
        $ cat << 'EOF' >> hello.php
        <?php
        function main(array $args) : array
        {
            echo "Some output to the invocation log";
            $name = $args["name"] ?? "stranger";
            $greeting = "Hello $name!";
            return ["greeting" => $greeting];
        }
        EOF
        $ wsk -i action update hello hello.php --kind php:7.1
        ok: updated action hello
        $ wsk -i action invoke hello -r -p name World

    If you see:

        {
            "greeting": "Hello World!"
        }

    Then it worked.
